README.md
# Tello Javascript

A minimalist Javascript library for programming Tello drones with a WebSocket interface for control, state and videoframes.

![Banner](./docs/tello-js-banner.jpg)

**Table of content**
* [How to Install](#how-to-install)
* [How to use](#how-to-use)
* [List of functions](#list-of-functions)
* [IP address and ports config](#ip-address-and-ports-config)
* [WebSocket Server for control](#websocket-server-for-control)
* [Command line interface (CLI)](#command-line-interface)
  * [tello cli tool](#tello-command)
	* [`tello start`](#tello-start)
	* [`tello set-wifi`](#telloset-wifi)
	* [`tello set-ap`](#telloset-ap)
  * [tello-server cli](#tello-server)
* [External Resources](#external-resources)

# How to install

ℹ️ To use this library you need to install:
- [Node.js](https://nodejs.org/)

1. Download/clone this repository on your computer.
```console
$ git clone https://github.com/harleylara/tello-js
```

2. In the **root directory** of **this** project execute.
```console
$ npm install -g .
```

Done. At this point you are ready to have fun with your drone using javascript.

# How to use

Basic usage:
```js
const Tello = require('tello.js')

const drone = Tello();

drone.connect(); // without arguments uses default ip and ports
drone.takeoff();
// cool stuff ...
drone.land();
drone.disconnect();
```

# List of functions

- async connect(tello_ip: str, control_port: int, state_port: int, video_port: int) {
- disconnect()
- getState() returns the latest state of the drone
- initFfmpeg() harley's note: this should be replaced with a function like `startVideo` that containt `streamon` and `initFfmpeg()`
- enableStream() for now is just sending the `streamon` command
- async sendCmd(command: str)
- async wait(time_ms: int)

# IP address and ports config

In case the drone is not using the default IP address and ports you can change the parameters in the `drone.json` file in the root directory of this repository.


# WebSocket Server for control

In case you want to control the drone using WebSocket you can configure the parameters in the `server.json` file in the root directory of this repository.

ℹ️ By default when the drone is initialized it broadcasts frames in BASE64 format to all clients connected to the default address `ws://localhost:3000`. You can change the address and port with the `drone.json` file in the `"videoServer"` section.

ℹ️ For more details: [WebSocket Server (Control, Set and State)](docs/socket.md)

# Command Line Interface

This package contain two cli tools to manage the tello drone:
- `tello` used launch tello-js and send command over the terminal
- `tello-server` launch tello-js with the websocket interface.

##  `tello` command

To get help on how to use this command:
```console
$ tello help

Usage: tello [options] [command]

CLI to manage Tello drone

Options:
  -V, --version       output the version number
  -h, --help          display help for command

Commands:
  start [options]     Start communication with the tello drone
  set-wifi [options]  Set Tello's WiFi network in station-mode (Tello create a WiFi network)
  set-ap [options]    Switch Tello into Access Point (AP) mode. Tello connect to an external WiFi network
  help [command]      display help for command
```
###  `tello start`

The command `tello start` launch tello-js in interactive mode, this allows you to send ℹ️ **RAW commands** ℹ️ to the drone directly from the terminal (commands from the SDK see [list of commands](./docs/socket.md#list-of-commands))

```console
$ tello start
```

To get documentation about to set flags use the command `help` before `start`

```console
$ tello help start

Usage: tello start [options]

Start communication with the tello drone

Options:
  --drone-ip <address>            tello drone IP address (default: "192.168.10.1")
  --control-port <number>         port to send control commands (default: 8889)
  --state-port <number>           port to get drone internal state (default: 8890)
  --video-port <number>           port to get video frames from drone (default: 11111)
  --video-socket-ip <address>     IP address serving video frames over websocket (default: "0.0.0.0")
  --video-socket-port <number>    port to get video frames over websocket (default: 3001)
  --control-socket-port <number>  port to send control commands over websocket (default: 3000)
  -h, --help                      display help for command
```

E.g. to launch `tello` with an especific IP address:
```sconsoleh
$ tello start --drone-ip 192.168.1.10
```

### `tello set-wifi`

The command `tello set-wifi` helps to setup drone's WiFi network in **station-mode**. ℹ️ Use the `-i` flag to do it interactively, the command terminal will ask for the name you want to give to the network and the password to be set.

```console
$ tello set-wifi -i
```

You can also configure the wifi in non-interactive mode using the appropriate flags.
```console
$ tello set-wifi -s NETWORK_NAME -p 123456789
```

To get documentation about to set flags user the command `help` before `set-wifi`
```console
$ tello help set-wifi

Usage: tello set-wifi [options]

Set Tello's WiFi network in station-mode (Tello create a WiFi network)

Options:
  -s, --ssid <name>      name of the drone's wifi network.
                            A prefix is always added, TELO-<your ssid>
                            Default 'TELLO-WIFI' (default: "WIFI")
  -p, --password <pass>  set password to connect drone's wifi. Default 123456789 (default: "123456789")
  -i, --interactive      set wifi in interactive mode
  -h, --help             display help for command

```

###  `tello set-ap`

The `tello set-ap` command does a similar job to `tello set-ap` but sets the drone's network into `access point mode`.

```console
$ tello set-ap -s TELLO-HOTSPOD -p 123456789
```

To get more informacion:
```console
$ tello help set-ap

Usage: tello set-ap [options]

Switch Tello into Access Point (AP) mode. Tello connect to an external WiFi network

Options:
  -s, --ssid <name>      Access Point name
  -p, --password <pass>  Access Point password
  -i, --interactive      set access-point mode in interactive mode
  -h, --help             display help for command

```

## `tello-server`

# External Resources
- [Tello SDK 2.0 - Official User Guide](https://dl-cdn.ryzerobotics.com/downloads/Tello/Tello%20SDK%202.0%20User%20Guide.pdf)
- [ROBOMASTER TT SDK 3.0 - Official User Guide](https://dl.djicdn.com/downloads/RoboMaster+TT/Tello_SDK_3.0_User_Guide_en.pdf)
- [Tello - User Guide](https://dl.djicdn.com/downloads/RoboMaster+TT/Tello_SDK_3.0_User_Guide_en.pdf)


HISTORY.md
# Tello JS

## In development:

- New Features:

- Notable Changes:

- Notable Fixes:

- Documentation Updates:

### 2023-02-12

- CLI added to the library with the command `tello`

### 2023-02-08

- File to start tello-core and tello-server `server.js` in the root folder was moved to `src` folder an renamed to `tello-server.js`

### 2022-10-29

- The `drone.json` and `server.json` were deleted and their contents combined into `config.json`.

### 2022-10-28

- A more reliable connection mechanism was implemented. Resends the connection command after a time without response.
- **BREAKING CHANGE**: `raw` keyword before a command string ex. `raw flip b` was removed. Now you can send SDK commands without `raw` in the begining of a string command ex. `flip b`

### 2022-10-23

- Logger module added to `tello.js`
- Ffmpeg is now automatically installed when the `npm install` command is executed.

### 2022-05-22

- Init server file for websocket use...

### 2022-05-19

- Logger added: contains 5 log categories (fatal, error, warning, information, success)


test/test.js
const wsServer = require('../src/server.js');
const Tello = require('../src/tello.js');

const drone = new Tello();
//drone.connect("192.168.137.97");
drone.connect();
// drone.sendCmd('takeoff');
drone.sendCmd('streamon');
drone.initFfmpeg();

const server = new wsServer(drone);


src/tello.js
const dgram = require('dgram');
const webSocket = require('ws');
const pathToFfmpeg = require('ffmpeg-static');
const child_process = require('child_process');
const stateDefinition = require('./stateDefinition.json');
const configs = require('../config.json');
const logger = require('./logger.js');

class Tello {
    /**
     * @constructor
     * @param {String} tello_ip Drone IP adress to bind to (control commands).
     * @param {Number} control_port UDP port to send control commands and receive answers
     * @param {Number} state_port UDP port to receive drone state
     */
    constructor() {

        // Control commands UDP Client
        this.controlClient = dgram.createSocket("udp4");

        // Drone State UDP Server
        this.stateServer = dgram.createSocket("udp4");

        // Drone State Object
        this.state = {};
        this.raw_state;
        this.stateDefinition = stateDefinition;

        this.ffmpegProcess;
        this.isStreaming = false;
        this.ffmpegRunnig = false;
        this.frame; // video frame

        this.connected = false;
        this.timeout = 500; // milliseconds
        this.max_time = 10 // seconds
        //this.executing = false;

    }

    /*
     * Initialize callbacks for the UDP control commnads client
     */
    initControlClient() {
        this.controlClient.on("message", (msg, rinfo) => {
            const response = msg.toString();
            logger.info(`Drone response: ${response} from ${rinfo.address}:${rinfo.port}`);

            if (!this.connected && response == 'ok') {
                // Initial connection
                this.connected = true;
            }
        });

        this.controlClient.on('error', (error) => {
            if (error) {
                logger.error(error);
                this.controlClient.close();
            }
        });

        this.controlClient.on("listening", () => {
            logger.info("Control Client Up and listening");
        });
    }

    /*
     * Initialize callbacks for the UDP drone state server
     */
    initStateServer() {
        this.stateServer.on("message", (msg, rinfo) => {
            const response = msg.toString().trim();// cleanup msg
            this.raw_state = response;
            this.state = this.parseState(response);
        });

        this.stateServer.on('error', (error) => {
            if (error) {
                logger.error(error);
            }
        });

        this.stateServer.on("listening", () => {
            logger.info("State Server Up and listening.");
        });
    }

    /*
     * Initialize callbacks for the WebSocket used for video streaming
     */
    initVideoSocket() {
        this.videoSocket.on('connection', (ws) => {
            console.log('New client connected to video socket.');

            // broadcast video stream to all clients
            ws.on('message', (msg) => {
                this.videoSocket.clients.forEach((client) => {
                    if (client.readyState === webSocket.OPEN) {
                        client.send(this.getFrame());
                    }
                });
            });
        });
        logger.info(`Video Socket Server started on ${this.VIDEO_SOCKET_IP}:${this.VIDEO_SOCKET_PORT}`);
    }

    /*
     * Delay helper in ms
     * @param {Number} time delay time in milliseconds
     */
    wait(time) {
        return new Promise(resolve => {
            setTimeout(() => {
                resolve(this);
            }, time);
        });
    }

    /**
     * Send a raw command (following the SDK format)
     * @param {String} command Command to send
     */
    sendCmd(cmd) {
        return new Promise((resolve, reject) => {
            const msg = Buffer.from(cmd);

            this.controlClient.send(msg, 0, msg.length, this.CONTROL_PORT, this.TELLO_IP, (error) => {
                if (error) {
                    logger.error("Error sending comand...");
                } else {
                    logger.info(`Command "${cmd}" sent to ${this.TELLO_IP}:${this.CONTROL_PORT}`);
                    resolve(this);
                }
            })
        })
    }

    /**
    * Initialize UDP connection with the Tello drone and Video Socket output
    * @param {tello_ip} drone IP
    * @param {control_port} port used to send control cmds
    * @param {state_port} port used to pull drone state variables
    * @param {video_port} video streaming port from Tello drone
    * @param {video_socket_ip} ip used to expose video frames
    * @param {video_socket_port} port used to provide video frame from drone
    */
    async connect(tello_ip, control_port, state_port, video_port, video_socket_ip, video_socket_port) {
        this.HOST = configs["host"] || "0.0.0.0";
        this.TELLO_IP = tello_ip || configs["drone"]["ip"] || "192.168.10.1";
        this.CONTROL_PORT = control_port || configs["drone"]["controlPort"] || 8889;
        this.STATE_PORT = state_port || configs["drone"]["statePort"] || 8890;
        this.VIDEO_PORT = video_port || configs["drone"]["videoPort"] || 11111;

        this.controlClient.bind(this.CONTROL_PORT);
        this.initControlClient();

        this.stateServer.bind(this.STATE_PORT);
        this.initStateServer();

        // Video Socket
        this.VIDEO_SOCKET_IP = video_socket_ip || configs["videoServer"]["ip"] || "0.0.0.0"
        this.VIDEO_SOCKET_PORT = video_socket_port || configs["videoServer"]["port"] || 3001
        this.videoSocket = new webSocket.Server({ host: this.VIDEO_SOCKET_IP, port: this.VIDEO_SOCKET_PORT });
        this.initVideoSocket();

        const msg = Buffer.from('command');
        logger.info(`Connecting to ${this.TELLO_IP} on port ${this.CONTROL_PORT}`);

        let attempts = 0;
        while (attempts < Math.round((this.max_time*1000)/this.timeout)) {
            
            await this.sendCmd(msg);
            await this.wait(this.timeout);

            if (this.connected){
                logger.success(`Connection succesfully with drone ${this.TELLO_IP}`)
                break
            }
            else {
                logger.error(`Fail to connect. retrying in ${this.timeout} ms...`)
                attempts++;
            }
        }
    }

    /**
     * Closes all sockets
     */
    disconnect() {
        this.controlClient.close();
        this.stateServer.close();
        this.ffmpegProcess.kill();
        this.connected = false;
    }

    /**
     * Parse raw state format into JSON object
     * @param {String} raw response from drone
     * @return JSON Objet with processed Drone State
     */
    parseState(raw_state) {
        // let test = "mid:-1;x:-100;y:-100;z:-100;mpry:0,0,0;pitch:8;roll:-2;yaw:69;vgx:0;vgy:0;vgz:-1;templ:75;temph:79;tof:10;h:0;bat:87;baro:-18.06;time:0;agx:116.00;agy:-4.00;agz:-1060.00;";
        let state = {}
        if (!!raw_state) {
            // NOTE: raw state end with ";" creating a empty element on split
            // spliting parameters and removing empty elements
            let state_parameters = raw_state.split(';').filter(param => !!param);

            // Processing parameters
            for (const param of state_parameters) {
                let key_value = param.split(':');
                if (key_value[0] !== "mpry") {
                    // replaces the original key with the "mapTo" key from stateDefinition file.
                    state[this.stateDefinition[key_value[0]]["mapTo"]] = { "value": parseFloat(key_value[1]), "unit": this.stateDefinition[key_value[0]]["unit"] };
                } else {
                    // Special condition for 'mpry' (mission pad orientation: roll pitch yaw)
                    // Split values and return array of float
                    let padOrientation = key_value[1].split(',').map(parseFloat);
                    state[this.stateDefinition[key_value[0]]["mapTo"]] = { "roll": padOrientation[0], "pitch": padOrientation[1], "yaw": padOrientation[2], "unit": this.stateDefinition[key_value[0]]["unit"] };
                }
            }
        }
        return state;
    }

    /*
     * @return JSON Object with Drone State
     */
    getState() {
        return this.state;
    }

    /*
     * Initialize Ffmpeg process to read UDP video stream from drone
     */
    initFfmpeg() {

        let videoPipe = {
            "command": pathToFfmpeg,
            "args": [
                '-re',
                '-y',
                '-i', `udp://${this.TELLO_IP}:${this.VIDEO_PORT}`,
                '-r', '30',
                '-s', '960x720',
                '-b', '800k',
                '-preset', 'ultrafast',
                '-tune', 'zerolatency',
                '-f', 'mjpeg',
                'pipe:1'
            ]
        }

        //TODO: before spawn check if the "streamon" command was already send
        this.ffmpegProcess = child_process.spawn(videoPipe['command'], videoPipe['args']);

        this.ffmpegProcess.on('error', (error) => {
            throw `Fail to spawn ffmpeg: ${error}`;
        });

        this.ffmpegProcess.on('close', (code) => {
            logger.error(`ffmpeg exited with code ${code}`);
        });

        this.ffmpegProcess.on('spawn', () => {
            this.ffmpegRunning = true;
            logger.success(`Ffmpeg spawned succesfully`);
        });

        this.ffmpegProcess.stderr.on('data', (data) => {
            // console.log(`stderr : ${data}`);
        });

        this.ffmpegProcess.stdout.on('data', (data) => {
            this.frame = `data:image/jpeg;base64,${Buffer.from(data).toString('base64')}`;
        });

    }

    getFrame() {
        return this.frame;
    }

    enableStream() {
        this.sendCmd('streamon');
    }
}

module.exports = Tello;


src/logger.js
const util = require("util");

class Logger {

    static stringify(...message) {
        const messages = [];
        if (message) {
            message.forEach(m => {
                if (m instanceof Object) {
                    try {
                        messages.push(JSON.stringify(m));
                    }
                    catch (error) {
                        messages.push(util.inspect(m, { color: true }));
                    }
                }
                else {
                    messages.push(m);
                }
            });
        }
        return messages;
    }

    static formating(log_type, ...message) {
        let now = new Date();
        let type = `[${log_type}]`
        return `${type.padEnd(10)},[${now.toUTCString()}],${this.stringify(message).join(' ')}`;
    }

    static fatal(...message) {
        // TODO add URL host:port to logs
        console.error(this.formating("FATAL", message));
        // process.exit(1);
    }

    static error(...message) {
        // TODO add URL host:port to logs
        console.error(this.formating("ERROR", message));
    }

    static warning(...message) {
        // TODO add URL host:port to logs
        console.error(this.formating("WARNING", message));
    }

    static info(...message) {
        // TODO add URL host:port to logs
        console.error(this.formating("INFO", message));
    }

    static success(...message) {
        // TODO add URL host:port to logs
        console.error(this.formating("SUCCESS", message));
    }
}

module.exports = Logger;


src/server.js
const WebSocket = require('ws');
const configs = require('../config.json');

class wsServer {

    /**
    * Initialize WebSocket server for send control commands
    * @param {control_port} port used to send control cmds TO the WebSocket Server
    *                       NOT to be confused with drone UDP socket port control.
    */
    constructor(DroneObj, control_port) {
        this.CONTROL_SERVER_IP = configs["controlServer"]["ip"] || "0.0.0.0";
        this.CONTROL_SERVER_PORT = control_port || configs["controlServer"]["port"] || 3000;

        // TODO: Check if the default port is in use and change to port++
        this.wss = new WebSocket.Server({ host: this.CONTROL_SERVER_IP, port: this.CONTROL_SERVER_PORT });
        this.drone = DroneObj;
        this.initWsServer();
    }

    initWsServer() {

        this.wss.on("connection", (ws) => {
            // TODO: create a list of clients ????
            // TODO: Allow only one client per connection ??
            console.log("New client connected to control socket");

            ws.on("message", (msg) => {
                console.log(`Client sent: ${msg}`);
                const fullMsg = msg.toString().trim().replaceAll(';', ' ');
                const command = fullMsg.split(' ')[0];
                console.log(`Full message: ${fullMsg}`)
                console.log(`Type of command: ${command}`);
                switch (command) {
                    case "takeoff":
                        ws.send("taking off");
                        this.drone.sendCmd("takeoff");
                        break;
                    case "land":
                        ws.send("landing");
                        this.drone.sendCmd("land");
                        break;
                    case "emergency":
                        ws.send("EMERGENCY something went wrong!!");
                        this.drone.sendCmd("emergency");
                        break;
                    case "state":
                        ws.send(JSON.stringify(this.drone.getState()));
                        break;
                    default:
                        // assume you are sending raw commands
                        // from the SDK
                        ws.send(`sending command: ${fullMsg}`)
                        this.drone.sendCmd(fullMsg);
                }
            })

            ws.on("close", () => {
                console.log("Client disconnected.")
            })
        })
    }
}

module.exports = wsServer;


src/cli.js
#!/usr/bin/env node
const wsServer = require('./server.js');
const Tello = require('./tello.js');
const package = require('../package.json');
const readline = require('readline');

const drone = new Tello();

/*
* Default settings
*/
const SSID = 'WIFI'; // station mode
const PASSWORD = '123456789'; // station mode

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
    prompt: 'cmd >> ',
    terminal: true
});

const { Command } = require('commander');
const { connected } = require('process');
const program = new Command();

/*
* Utility function to show information about the cli
* TODO?: info about the OS, node version??
*/
function showInfo(){
    console.log("-----------------------------------")
    console.log("Tello interactive CLI started");
    console.log(`tello-js version: ${package.version}`);
    console.log("-----------------------------------")
}

/*
* Utility function for interactive mode
* @param {String} question text
*/
function ask(questionText) {
    return new Promise((resolve, reject) => {
        rl.question(questionText, (input) => resolve(input) );
    });
}

async function start() {
    const options = this.opts();

    showInfo();

    console.log(`Drone IP: ${options.droneIp}`);
    console.log(`Control port: ${options.controlPort}`)
    console.log(`State port: ${options.statePort}`)
    console.log(`Video port: ${options.videoPort}`)
    console.log(`Video socket ip: ${options.videoSocketIp}`);
    console.log(`Video socket port: ${options.videoSocketPort}`);
    console.log(`Control socket port: ${options.controlSocketPort}`)
    console.log("-----------------------------------")

    await drone.connect(
        tello_ip        = options.telloIp,
        control_port    = options.controlPort,
        state_port      = options.statePort,
        video_port      = options.videoPort,
        video_socket_ip = options.videoSocketIp,
        video_socket_port = options.videoSocketPort

    )

    if (drone.connected){

        const server = new wsServer(drone, 
            control_port = options.controlSocketPort
        );

        // drone.sendCmd('streamon');
        // drone.initFfmpeg();

        rl.on("line", cliInput);
    } else {
        console.log("FAIL to connect");
    }
}


async function cliInput(input){
    if (input === "exit"){
        rl.close();
    }
    await drone.sendCmd(input);
    // rl.prompt();
}


async function setWifi(){
    const options = this.opts();

    let ssid = options.ssid;
    let password = options.password;

    showInfo();

    await drone.connect();

    if (drone.connected){

        if (options.interactive === undefined){
            console.log("Set WiFi station-mode started in non-interactive")
        } else {
            console.log("Set WiFi station-mode started in interactive mode")
            ssid = await ask('Set WiFi SSID: ');
            password = await ask('Set password: ');
        }

        console.log(`WiFi SSID: ${ssid}`);
        console.log(`WiFi password: ${password}`);

        await drone.sendCmd(`wifi ${ssid} ${password}`);
        console.log("The drone will automatically reboot in 3s");
        process.exit(0);

    } else {
        console.log("FAIL to connected");
        process.exit(1);
    }
}

async function setAP(){
    const options = this.opts();

    let ssid = options.ssid;
    let password = options.password;

    showInfo();

    // await drone.connect();

    if (drone.connected){

        if (options.interactive === undefined){
            console.log("Set access-point-mode started in non-interactive")
        } else {
            console.log("Set access-point-mode started in interactive mode")
            ssid = await ask('Set Access Point SSID: ');
            password = await ask('Set Access Point password: ');
        }

        console.log(`Access Point SSID: ${ssid}`);
        console.log(`Access Point password: ${password}`);

        await drone.sendCmd(`ap ${ssid} ${password}`);
        console.log("The drone will automatically reboot in 3 seconds");
        process.exit(0);

    } else {
        console.log("FAIL to connected");
        process.exit(1);
    }

}

async function main() {
  program
    program
        .name('tello')
        .description('CLI to manage Tello drone')
        .version(package.version);

    program
        .command('start')
        .description('Start communication with the tello drone')
        .option('--drone-ip <address>', 'tello drone IP address', '192.168.10.1')
        .option('--control-port <number>', 'port to send control commands', 8889)
        .option('--state-port <number>', 'port to get drone internal state', 8890)
        .option('--video-port <number>', 'port to get video frames from drone', 11111)
        .option('--video-socket-ip <address>', 'IP address serving video frames over websocket', '0.0.0.0')
        .option('--video-socket-port <number>',  'port to get video frames over websocket', 3000)
        .option('--control-socket-port <number>', 'port to send control commands over websocket', 3001)
        .action(start)

    program
        .command('set-wifi')
        .description("Set Tello's WiFi network in station-mode (Tello create a WiFi network)")
        .option('-s, --ssid <name>', `name of the drone's wifi network.
                            A prefix is always added, TELO-<your ssid>
                            Default 'TELLO-${SSID}'`, SSID)
        .option('-p, --password <pass>', `set password to connect drone's wifi. Default ${PASSWORD}`, PASSWORD)
        .option('-i, --interactive', "set wifi in interactive mode")
        .action(setWifi)

    program
        .command('set-ap')
        .description("Switch Tello into Access Point (AP) mode. Tello connect to an external WiFi network")
        .option('-s, --ssid <name>', `Access Point name`)
        .option('-p, --password <pass>', `Access Point password`)
        .option('-i, --interactive', "set access-point mode in interactive mode")
        .action(setAP)

    await program.parseAsync(process.argv);
}

main();
